src/main/java/dev/openfeature/sdk/ProviderRepository.java

```diff
@@ -63,66 +64,70 @@ public void setProvider(String clientName, FeatureProvider provider) {
         initializeProvider(clientName, provider);
     }
 
-    private void initializeProvider(FeatureProvider provider) {
-        initializingDefaultProvider = provider;
-        initializeProvider(provider, this::updateDefaultProviderAfterInitialization);
-    }
-
-    private void initializeProvider(String clientName, FeatureProvider provider) {
-        initializingNamedProviders.put(clientName, provider);
-        initializeProvider(provider, newProvider -> updateProviderAfterInit(clientName, newProvider));
-    }
-
-    private void initializeProvider(FeatureProvider provider, Consumer<FeatureProvider> afterInitialization) {
+    /**
+     * Get the emitter for the referenced clientName, or the default.
+     * The emitter is created if it doesn't exist.
+     * 
+     * @param clientName name for the client, or null for default.
+     * @return existing or new emitter for this clientName (or the default).
+     */
+    EventEmitter getAndCacheEmitter(@Nullable String clientName) {
+        return Optional
+            .ofNullable(clientName)
+            .map(name -> {
+                return Optional
+                    .ofNullable(this.clientEmitters.get(name))
+                    .filter(emitter -> Optional.ofNullable(emitter).isPresent())
+                    .orElseGet(() -> {
+                        EventEmitter newEmitter = new EventEmitter();
+                        this.clientEmitters.put(name, newEmitter);
+                        return newEmitter;
+                    });
+            })
+            .orElse(defaultEmitter);
+    }
+
+    private void initializeProvider(@Nullable String clientName, FeatureProvider newProvider) {
+        forwardProviderEvents(clientName, newProvider);
         taskExecutor.submit(() -> {
             try {
-                if (!isProviderRegistered(provider)) {
-                    provider.initialize();
+                if (!isProviderRegistered(newProvider)) {
+                    newProvider.initialize(OpenFeatureAPI.getInstance().getEvaluationContext());
```

Comment:
Is the Global EvaluationContext set in the OpenFeature API common to all providers? Is there any way a specific provider can be initialized with context specific to that provider?

Reply:
> Is the Global EvaluationContext set in the OpenFeature API common to all providers? Is there any way a specific provider can be initialized with context specific to that provider?

This is an interesting though, but not something we have spec'd. It's the same global context for each provider at init. The context set at the `client` level would be specific to the provider - but that wouldn't be available at init.